wizard.py

```diff
@@ -0,0 +1,1038 @@
+
+  """
+  steps = []
+  for i in range(len(step_names)):
+    steps.append({
+        "name": step_names[i],
+        "cmd": step_commands[i],
+        "modifies": step_modifies[i] == "true"
+      })
+
+  inspections = []
+  for i in range(len(inspection_names)):
+    inspections.append({
+        "name": inspection_names[i],
+        "cmd": inspection_commands[i],
+        "based_on": inspection_step_names[i]
+      })
+
+  ssc_data = {
+    "steps": steps,
+    "inspections": inspections
+  }
+  return ssc_data
+
+
+def _auth_items_to_dict(auth_items):
+  """Takes a list of auth_items and returns a dictionary mapping the items
+  to their respective step names, i.e..:
+  [{"step_name": ..., "threshold": ..., "authorized_functionaries": ...}, ...]
+  -->
+  {<step name> : {"step_name": ..., "threshold": ..., "authorized_..."}, ...}
+  """
+  auth_dict = {}
+  for auth_item in auth_items:
+    auth_dict[auth_item["step_name"]] = auth_item
+  return auth_dict
+
+
+# -----------------------------------------------------------------------------
+# NoSQL Helpers
+# -----------------------------------------------------------------------------
+# NOTE: Below functions rely on the current session having an id. If there is
+# no id in the session, all functions redirect to `404` (page not found).
+# This should never happen because all calling views should be decorated with
+# @with_session_id, which ensures that the current session does have an id.
+def _persist_session_subdocument(subdocument):
+  """Update a subdocument (e.g. vcs, ssc, functionaries...) in session document
+  identified by current session id. """
+  if not session.get("id"):
+    abort(404)
+
+  # Search session document by session ID in DB and update (replace)
+  # subdocument. If the entire document does not exist it is inserted
+  mongo.db.session_collection.update_one(
+    {"_id": session["id"]},
+    {"$set": subdocument},
+    upsert=True)
+
+
+def _persist_session_subdocument_ts(subdocument):
+  """Updates/adds last_modified to the subdocument before persisting it. """
+
+  for key in subdocument.keys():
+    subdocument[key]["last_modified"] = time.time()
+  _persist_session_subdocument(subdocument)
+
+
+def _get_session_subdocument(key):
+  """Returns a subdocument (e.g. vcs, ssc, functionaries...) identified by
+  passed key from session document identified by current session id.
+  Returns an empty dict if document or subdocument are not found.  """
+  if not session.get("id"):
+    abort(404)
+
+    # Get session document (use short circuit for default empty dict)
+  session_doc = mongo.db.session_collection.find_one(
+      {"_id": session["id"]})
+
+  if not session_doc:
+    return {}
+
+  # Get vcs data from session document or empty dict
+  return session_doc.get(key, {})
+
+
+def _get_session_document():
+  """Returns the entire session document.
+  Returns an empty dict if document or subdocument are not found.  """
+
+  if not session.get("id"):
+    abort(404)
```

Comment:
I'm starting to think these HTTP error codes are somewhere in flask. If not, they *should* be.

Reply:
Somewhere in Flask? As constant?

Reply:
Apparently, it doesn't have these defined, but abort() searches for an exception with this code. Do you think it's worth defining? I'm thinking of ease of refactoring in case we want to change the error code.

Reply:
I'm not convinced. To me `404` is semantically strong. Let's leave it for now and think about it if it ever becomes a problem.